File: othello_python/player.py
"""
Class Player
Defines one player for Othello
Akseli Lukkarila
2019
"""
import random
import time

from typing import List

from board import Board
from colorprint import Color, print_color, print_error
from utils import Disk, Move, Square


class Player:
    """Defines one player."""
    def __init__(self, disk: Disk, human=True, show_helpers=True):
        self._can_play = True
        self._disk = disk
        self._human = human
        self._rounds_played = 0
        self._show_helpers = show_helpers

    def can_play(self) -> bool:
        return self._can_play

    def play_one_move(self, board: Board) -> None:
        """Play one round as this player."""
        print(f"Turn: {str(self._disk)}")
        moves = board.possible_moves(self._disk)
        if moves:
            self._can_play = True
            if self._human and self._show_helpers:
                board.print_moves(moves)

            chosen_move = self._get_human_move(moves) if self._human else self._get_computer_move(moves)
            board.place_disk(chosen_move)
            board.print_score()
            self._rounds_played += 1
        else:
            self._can_play = False
            print_color("  No moves available...", Color.yellow)

    def set_human(self, is_human: bool):
        """Set player to be controlled by human or computer."""
        self._human = is_human

    @staticmethod
    def _get_computer_move(moves: List[Move]) -> Move:
        """Return move chosen by computer."""
        print("  Computer plays...")
        # wait a bit and pick a random move
        time.sleep(random.uniform(1.0, 2.0))
        move = random.choice(moves)
        print(f"  -> {move.square}")
        return move

    def _get_human_move(self, moves: List[Move]) -> Move:
        """Return move chosen by a human player."""
        while True:
            square = self._get_square()
            # check if given square is one of the possible moves
            move = next((move for move in moves if move.square == square), None)
            if move:
                return move

            print_error(f"can't place a {str(self._disk)} disk in square {square}!", indent=2)

    @staticmethod
    def _get_square() -> Square:
        """Ask human player for square coordinates."""
        while True:
            try:
                pos = input("  Give disk position (x,y): ")
                coordinates = [int(x.strip()) for x in pos.split(",") if x.strip()]
                if len(coordinates) != 2:
                    raise ValueError
                return Square(coordinates[0], coordinates[1])
            except ValueError:
                print_error("give coordinates in the form 'x,y'!", indent=2)

    def _type_string(self) -> str:
        return "Human   " if self._human else "Computer"

    def __str__(self):
        return f"{str(self._disk)} | {self._type_string()} | Moves: {self._rounds_played}"

File: othello_csharp/Othello/Player.cs
//==========================================================
// Class Player
// Defines one player for Othello
// Akseli Lukkarila
// 2019-2023
//==========================================================

using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Threading;

namespace Othello
{
    /// Defines one player (human or computer).
    internal class Player
    {
        public bool canPlay;
        private bool _isHuman;
        private int _roundsPlayed;
        private readonly Disk _disk;
        private readonly Random _random;
        private readonly bool _showHelpers;

        public Player(Disk color)
        {
            _disk = color;
            _isHuman = true;
            _random = new Random();
            _showHelpers = true;
            canPlay = true;
        }

        /// Play one round as this player.
        public void PlayOneMove(Board board)
        {
            Console.WriteLine($"Turn: {_disk.Name()}");
            var moves = board.PossibleMoves(_disk);
            if (moves.Any())
            {
                canPlay = true;
                if (_isHuman && _showHelpers)
                {
                    board.PrintMoves(moves);
                }
                var chosenMove = _isHuman ? GetHumanMove(moves) : GetComputerMove(moves);
                board.PlaceDisc(chosenMove);
                board.PrintScore();
                ++_roundsPlayed;
                Thread.Sleep(1000);
            }
            else
            {
                canPlay = false;
                ColorPrint.WriteLine("  No moves available...", Color.Yellow);
            }
        }

        /// Set player to be controlled by human or computer.
        public void SetHuman(bool isHuman)
        {
            _isHuman = isHuman;
        }

        /// Reset player status for a new game.
        public void Reset()
        {
            canPlay = true;
            _roundsPlayed = 0;
        }

        /// Return move chosen by computer.
        private Move GetComputerMove(IReadOnlyList<Move> moves)
        {
            Console.WriteLine("  Computer plays...");
            // Wait a bit and pick a random move
            Thread.Sleep(_random.Next(1000, 2000));
            var chosenMove = moves[_random.Next(moves.Count)];
            Console.WriteLine($"  {chosenMove.Square} -> {chosenMove.Value}");
            return chosenMove;
        }

        /// Return move chosen by a human player.
        private Move GetHumanMove(List<Move> moves)
        {
            while (true)
            {
                var square = GetSquare();
                // check if given square is one of the possible moves
                if (moves.Exists(x => square.Equals(x.Square)))
                {
                    return moves.Find(x => square.Equals(x.Square));
                }
                ColorPrint.Error($"  Can't place a {_disk.Name()} disk in square {square}!");
            }
        }

        /// Ask human player for square coordinates.
        private static Square GetSquare()
        {
            while (true)
            {
                try
                {
                    Console.Write("  Give disk position (x,y): ");
                    var coords = Console.ReadLine();
                    if (string.IsNullOrEmpty(coords) || coords.Length != 3 || coords[1] != ',')
                    {
                        throw new FormatException("Invalid coordinates");
                    }
                    var x = int.Parse(coords[0..1]);
                    var y = int.Parse(coords[2..3]);
                    return new Square(x, y);
                }
                catch (FormatException)
                {
                    ColorPrint.Error("  Give coordinates in the form 'x,y'");
                }
            }
        }

        /// Return player type description string.
        private string TypeString()
        {
            return _isHuman ? "Human   " : "Computer";
        }

        public override string ToString()
        {
            return $"{_disk.Name()} | {TypeString()} | Moves: {_roundsPlayed}";
        }
    }
}
